src/plugins/ssh/index.ts

```diff
@@ -0,0 +1,404 @@
+ * are printed to the error output within the first 5 seconds of startup.
+ * If they are, the returned `isAccessPropagated()` function will return false.
+ * When this occurs, the consumer of this function should retry the `aws` / `gcloud` command.
+ *
+ * Note that this function requires interception of the subprocess stderr stream.
+ * This works because AWS SSM wraps the session in a single-stream pty, so we
+ * do not capture stderr emitted from the wrapped shell session.
+ */
+const accessPropagationGuard = (
+  child: ChildProcessByStdio<null, null, Readable>,
+  debug?: boolean
+) => {
+  let isEphemeralAccessDeniedException = false;
+  let isGoogleLoginException = false;
+  const beforeStart = Date.now();
+
+  child.stderr.on("data", (chunk) => {
+    const chunkString: string = chunk.toString("utf-8");
+
+    if (debug) print2(chunkString);
+
+    const match = UNPROVISIONED_ACCESS_MESSAGES.find((message) =>
+      chunkString.match(message.pattern)
+    );
+
+    if (
+      match &&
+      Date.now() <=
+        beforeStart + (match.validationWindowMs || DEFAULT_VALIDATION_WINDOW_MS)
+    ) {
+      isEphemeralAccessDeniedException = true;
+    }
+
+    const googleLoginMatch = chunkString.match(GOOGLE_LOGIN_MESSAGE);
+    isGoogleLoginException = isGoogleLoginException || !!googleLoginMatch; // once true, always true
+    if (isGoogleLoginException) {
+      isEphemeralAccessDeniedException = false; // always overwrite to false so we don't retry the access
+    }
+  });
+
+  return {
+    isAccessPropagated: () => !isEphemeralAccessDeniedException,
+    isGoogleLoginException: () => isGoogleLoginException,
+  };
+};
+
+const spawnChildProcess = (
+  credential: AwsCredentials | undefined,
+  command: string,
+  args: string[],
+  stdio: [StdioNull, StdioNull, StdioPipe]
+) =>
+  spawn(command, args, {
+    env: {
+      ...process.env,
+      ...credential,
+    },
+    stdio,
+    shell: false,
+  });
+
+type SpawnSshNodeOptions = {
+  credential?: AwsCredentials;
+  command: string;
+  args: string[];
+  attemptsRemaining: number;
+  abortController?: AbortController;
+  detached?: boolean;
+  stdio: [StdioNull, StdioNull, StdioPipe];
+  provider: "aws" | "gcloud";
+  debug?: boolean;
+};
+
+const friendlyProvider = (provider: "aws" | "gcloud") =>
+  provider === "aws"
+    ? "AWS"
+    : provider === "gcloud"
+      ? "Google Cloud"
+      : throwAssertNever(provider);
+
+/** Starts an SSM session in the terminal by spawning `aws ssm` as a subprocess
+ *
+ * Requires `aws ssm` to be installed on the client machine.
+ */
+
+async function spawnSshNode(
+  options: SpawnSshNodeOptions
+): Promise<number | null> {
+  return new Promise((resolve, reject) => {
+    const child = spawnChildProcess(
+      options.credential,
+      options.command,
+      options.args,
+      options.stdio
+    );
+
+    // TODO ENG-2284 support login with Google Cloud: currently return a boolean to indicate if the exception was a Google login error.
```

Comment:
does this still apply

Reply:
Yes, it's still a TODO to automatically login (or at least prompt a second login)